File: Go/039. Combination Sum.go
package main

import "sort"

func combinationSum(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	return findSum(candidates, target)
}

func findSum(candidates []int, target int) [][]int {
	res := make([][]int, 0, 0)
	if len(candidates) == 0 || target < candidates[0] {
		return res
	}
	if candidates[0] == target {
		tmp := []int{target}
		res = append(res, tmp)
		return res
	}
	tmp1 := findSum(candidates[1 :], target)
	tmp2 := findSum(candidates, target - candidates[0])
	for i := 0; i < len(tmp2); i++ {
		tmp2[i] = append(tmp2[i], candidates[0])
	}
	res = append(res, tmp1...)
	res = append(res, tmp2...)
	return res
}

File: C++/039. Combination Sum.cpp
#include <iostream>
#include <algorithm>
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> combinationSum(std::vector<int>& candidates, int target) {
        std::sort(candidates.begin(), candidates.end());
        return findSum(candidates, target);
    }

    std::vector<std::vector<int>> findSum(std::vector<int> candidates, int target) {
        std::vector<std::vector<int>> res;
        if (candidates.size() == 0 || target < candidates[0])
            return res;
        if (candidates[0] == target) {
            res.emplace_back(std::vector<int>{target});
            return res;
        }
        std::vector<std::vector<int>> tmp1 = findSum(std::vector<int>(candidates.begin() + 1, candidates.end()), target);
        std::vector<std::vector<int>> tmp2 = findSum(candidates, target - candidates[0]);
        for (auto &e : tmp2) {
            e.insert(e.begin(), candidates[0]);
        }
        res.insert(res.end(), tmp1.begin(), tmp1.end());
        res.insert(res.end(), tmp2.begin(), tmp2.end());
        return res;
    }
};